tests/infra/proposal_generator.py

```diff
@@ -0,0 +1,334 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache 2.0 License.
+
+import argparse
+import inspect
+import json
+import os
+
+from loguru import logger as LOG
+
+
+def dump_proposal(output_path, proposal, dump_args):
+    LOG.info(f"Writing proposal to {output_path}")
+    with open(output_path, "w") as f:
+        json.dump(proposal, f, **dump_args)
+
+
+def dump_vote(output_path, vote, dump_args):
+    LOG.info(f"Writing vote to {output_path}")
+    with open(output_path, "w") as f:
+        json.dump(vote, f, **dump_args)
+
+
+def file_to_byte_array(path):
+    with open(path, "r") as f:
+        return [ord(c) for c in f.read()]
+
+
+def list_as_lua_literal(l):
+    return str(l).translate(str.maketrans("[]", "{}"))
+
+
+def script_to_vote_object(script):
+    return {"ballot": {"text": script}, "id": PROPOSAL_ID_PLACEHOLDER}
+
+
+TRIVIAL_YES_BALLOT = {"text": "return true"}
+TRIVIAL_NO_BALLOT = {"text": "return false"}
+
+LUA_FUNCTION_EQUAL_ARRAYS = """
+function equal_arrays(a, b)
+  if #a ~= #b then
+    return false
+  else
+    for k, v in ipairs(a) do
+      if b[k] ~= v then
+        return false
+      end
+    end
+    return true
+  end
+end
+"""
+
+
+PROPOSAL_ID_PLACEHOLDER = "<replace with desired proposal_id>"
+
+
+def add_arg_construction(lines, arg, arg_name="args"):
+    if isinstance(arg, list):
```

Comment:
In general it's a bit nicer to take collections.abc.Sequence, Mapping (with a special case for basestring, which is a Sequence but needs different handling). It's a more precise requirement, and it means passing generators, custom iterators etc would work.